rcore/log.c

```diff
@@ -34,58 +35,91 @@ void log_printf_to_ar(const char *layer, const char *module, uint8_t level, cons
 
 void log_printf(const char *layer, const char *module, uint8_t level, const char *filename, uint32_t line_no, const char *fmt, va_list ar)
 {
-    char log_type[12];
-    char buf[16];
+    uint8_t interrupt_set = 0;
+    static BaseType_t xHigherPriorityTaskWoken;
+    char buf[128];
+    char tbuf[16];
     
     // XXX: this means that log_printf *must* be called first from a non-threaded context, for this check is not thread-safe!
     // XXX: verify this here.
     if (_log_mutex == NULL)
         _log_mutex = xSemaphoreCreateMutexStatic(&_log_mutex_buf);
     
-    xSemaphoreTake(_log_mutex, portMAX_DELAY);
+    if(is_interrupt_set())
+    {
+        xHigherPriorityTaskWoken = pdFALSE;
+        xSemaphoreTakeFromISR(_log_mutex, &xHigherPriorityTaskWoken);
```

Comment:
hmm, is this safe?  are ISRs allowed to block?  (how does that even work?)

Reply:
Should be safe. We interrogate the ICSR VECTACTIVE mask to see if we in an ISR. If we detect an ISR call, we use the _FromISR versions from FreeRTOS. They are non-blocking, so in fact the call fails. Should likely detect this failure and do something about it, but it isn't critical, and in fact there are few calls to logging from ISRs

Reply:
Should be safe, yeah. We interrogate the ICSR VECTACTIVE mask to see if there the call is from an ISR. Then we call the _FromISR  version of the function. This is non-blocking (that would be bad mmkay). The bug I see here if that the failure mode of getting the semaphore isn't handled. In this case, it isn't a biggy. we silently fail and get some text thLogEntry!at may be b0rked. XXX fixme TODO I suppose